### PR TITLE
add scatter gather configs

### PR DESCRIPTION
**Pull request by lmbollen**
This PR goes on top of #66
 
This PR adds GADTs to configure the scatter- and gather units. 
These GADTs hide the component specific type variables from the rest of the implementation.

### CODE REVIEW

File: bittide/src/Bittide/ScatterGather.hs [hunk 1]
```diff
@@ -31,7 +47,7 @@ scatterUnit ::
   , KnownNat memDepth, 1 <= memDepth
   , KnownNat frameWidth) =>
   -- | Configuration for the 'calendar'.
-  CalendarConfig nBytes addrW (CalendarEntry memDepth) ->
+  CalendarConfig nBytes addrW (Index memDepth) ->
```
Comment (leonschoorl) [inline bittide/src/Bittide/ScatterGather.hs:50]: Shouldn't this take a `ScatterConfig`?

File: bittide/src/Bittide/ScatterGather.hs [hunk 2]
```diff
@@ -59,7 +75,7 @@ gatherUnit ::
   , KnownNat frameWidth, 1 <= frameWidth
   , KnownNat (DivRU frameWidth 8), 1 <= (DivRU frameWidth 8)) =>
   -- | Configuration for the 'calendar'.
-  CalendarConfig nBytes addrW (CalendarEntry memDepth) ->
+  CalendarConfig nBytes addrW (Index memDepth) ->
```
Comment (leonschoorl) [inline bittide/src/Bittide/ScatterGather.hs:78]: Shouldn't this take the `GatherConfig` instead?

**Review by leonschoorl — APPROVED** [review]
Otherwise LGTM

File: bittide/src/Bittide/ScatterGather.hs (hunk 1 above)
Comment (lmbollen) [reply bittide/src/Bittide/ScatterGather.hs:50]: The `GADT`s are used to hide implementations details where they are not necessary and be able to differentiate  between the S/G unit calendars.
The differentiation would be nice to have, but `memDepth` is present in the API of the `scatterUnit` and gatherUnit`. I could change it, but I'd have to add a run-time check that verifies whether the tyVars are the same.

File: bittide/src/Bittide/ScatterGather.hs [hunk 3]
```diff
@@ -22,12 +22,16 @@ import Bittide.Calendar
 import Bittide.DoubleBufferedRam
 import Bittide.SharedTypes
 
+-- | GADT to explicitly differentiate between a configuration for the 'scatterUnitWb' and
```
Comment (leonschoorl) [inline bittide/src/Bittide/ScatterGather.hs:25]: This is actually a GADT, but an existential type, that just happens to be defined using [GADTSyntax](https://downloads.haskell.org/ghc/9.2.4/docs/html/users_guide/exts/gadt_syntax.html#extension-GADTSyntax)

Comment (leonschoorl) [reply bittide/src/Bittide/ScatterGather.hs:25]: But you can probably leave those details out, if what you care about documenting here is that they allow you differentiate between scatter and gather config.
The existential part is what allows you to hide the `memDepth`.

File: bittide/src/Bittide/ScatterGather.hs (hunk 1 above)
Comment (leonschoorl) [reply bittide/src/Bittide/ScatterGather.hs:50]: But you're right, not having access to the `memDepth` inside the `ScatterConfig` would be problematic here

**Pull request merged**